[WIP]: Don't produce an output file and init procs for an empty module

cooldome opened the pull request:
Implements  language feature #9798
Not ready yet

cooldome commented:
Hi, Is anything funny going on with the testament lately?
I can't reproduce any of the reported test failures locally. 
Error messages are such that it should be straightforward to reproduce.
I did get some other failures and those I have fixed.

Any ideas?

Araq commented:
Try the new "megatest" test category.

Araq commented:
This seems to be a complex rewrite of the logic. Have you considered enabling LTO instead? ;-)

Alternatively we could introduce some postprocessing step in the codegen to cleanup some things.

cooldome commented:
In this PR I am more concerned with fact that 40% of modules after codegen are empty in my project, yet cpp files are produced and cpp compiler is invoked and it takes significant amount of compilation  time to run cpp compiler. Dead code elimination for entire modules and their init procs is very useful feature.

The logic rewrite is actually small and and new logic is not more complicated then existing one in my opinion. it is more of `git diff` failure to identify what parts are actually changed and actually a lot of changes. I can give you a  summary of what has changed:

1. `genInitProc` is split into two procs: genInitCode and genDatInitCode so can eliminate `init` and `datInit` procs independently.
2. `genMainProc` now needs to be executed after all `genInitCode` calls have been made in order to identify which modules have init procs. Because `genInitCode` is done in `cgenWriteModules` I had make sure that modules written in a specific order to generate `init` proc invocations in the correct order. `genMainProc` had some `cgsym` calls and it is too late to do it in `cgenWriteModules`, so I had to move `cgsyms` calls to `myClose`.
3. System module is no longer treated specially, the system module and/or its init procs can be eliminated as any other modules. In test with `os:standalone` and `gc: none` I check that both `systemInit` and `systemDatInit` are eliminated. I had changed definition of `nimvm` in system.nim to compileTime only as it was the only thing that prevented dce of `systemInit`.
4. Previously, init calls where made int this order: systemDatInit, init TLS, init GC, calls other modules dataInit, call all modules init. In order to achieve item `3`, I have changed  it to:
 calls all modules datInit, init TLS, init GC, call all modules init

cooldome commented:
Ready for review

cooldome commented:
I have updated the summary above

cooldome commented on compiler/cgen.nim:
```diff
@@ -1290,82 +1314,115 @@ proc genInitCode(m: BModule) =
     # Keep a bogus frame in case the code needs one
     add(prc, ~"\tTFrame FR_; FR_.len = 0;$N")
 
+    if m.preInitProc.s(cpsLocals).len > 0:
```
git diff algorithm failure here: there is no new lines here, it is existing lines wrapped into `if rope.len > 0: initRequired = true`

cooldome commented on compiler/cgen.nim:
```diff
@@ -1290,82 +1314,115 @@ proc genInitCode(m: BModule) =
+    if optStackTrace in m.initProc.options and preventStackTrace notin m.flags:
+      add(prc, deinitFrame(m.initProc))
 
-  if optStackTrace in m.initProc.options and frameDeclared notin m.flags:
```
git diff algorithm failure here: no lines are deleted in genInitProcCode

cooldome commented on compiler/cgen.nim:
```diff
@@ -1260,22 +1256,50 @@ proc getInitName(m: PSym): Rope =
+  if moduleDatInitRequired:
+    add(m.s[cfsDatInitProc], prc)
+
 proc genInitCode(m: BModule) =
```
git diff messed up the changes here completely. It shows that many lines are added and many lines deleted while actually no lines were deleted in `genInitCode` proc and only a couple were added. No changes to init logic were made, only code wrapped into:
    if rope.len > 0:
      moduleInitRequired = true
